ecommerce/models.py

```diff
@@ -223,6 +235,85 @@ class Coupon(Model):
     )
     enabled = BooleanField(default=True, help_text="If true, coupons are presently redeemable")
 
+    @property
+    def program(self):
+        """Get the program from the content_type"""
+        obj = self.content_object
+        if isinstance(obj, Program):
+            return obj
+        elif isinstance(obj, Course):
+            return obj.program
+        else:
+            # This must be a CourseRun. This is validated in clean()
+            return obj.course.program
+
+    @property
+    def course_keys(self):
+        """Get the course keys which are targets of this coupon"""
+        obj = self.content_object
+        if isinstance(obj, Program):
+            return CourseRun.objects.filter(course__program=obj).values_list('edx_course_key', flat=True)
+        elif isinstance(obj, Course):
+            return CourseRun.objects.filter(course=obj).values_list('edx_course_key', flat=True)
+        else:
+            # This must be a CourseRun. This is validated in clean()
+            return [obj.edx_course_key]
+
+    @property
+    def is_valid(self):
+        """Returns true if the coupon is enabled and also within the valid date range"""
+        now = datetime.now(tz=pytz.UTC)
+        if not self.enabled:
+            return False
+        if self.activation_date is not None and now < self.activation_date:
+            return False
+        if self.expiration_date is not None and now > self.expiration_date:
+            return False
+        return True
+
+    @property
+    def is_automatic(self):
+        """
+        Returns true if the coupon would be redeemed automatically without input from the user.
+        """
+        return self.coupon_type == self.DISCOUNTED_PREVIOUS_RUN
+
+    def user_has_redemptions_left(self, user):
+        """
+        Has the coupon already been redeemed the max number of times?
+        """
+        return RedeemedCoupon.objects.filter(order__user=user, coupon=self).count() < self.num_redemptions_per_user
+
+    def is_redeemable(self, user, course_key):
+        """
+        Returns true if the coupon is redeemable for the user.
+        """
+        if (
+                not self.is_valid or                      # coupon must be enabled and within valid date range
+                course_key not in self.course_keys or     # coupon must apply to a course run we care about
+                self.num_coupons_available == 0 or        # coupon must have a redemption available
+                not self.user_has_redemptions_left(user)  # coupon must not be used up for that user
+        ):
+            return False
```

Comment:
Reading through the logic in this method, it looks like unless conditions are met for this `False` return, all other code paths return `True` so the logic below is fairly moot (i.e., might as well return here). Is this future-proofing or did I miss something?